avocado/core/defaults.py

```diff
@@ -0,0 +1,20 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2018
+# Author: Cleber Rosa <crosa@redhat.com>
+
+"""
+The Avocado core defaults
+"""
+
+#: The encoding used by default on all data input
+ENCODING = 'utf-8'
```

Comment:
I don't think we should add yet another source of information. We have `settings` for global configuration and `varianter` for test-related parameters.

Reply:
I obviously think differently.  Settings is one thing, hard coded, builtin defaults are a lower level.  On other code I've written, which I considered a successful approach, I had settings (and even command line options, IIRC) all reading from defaults to set their own *defaults*.

Reply:
I don't say we can't use `defaults`, but adding such module is on a longer discussion and should incorporate all already existing defaults (should we decide to go this way) and not just added ad-hock for a single purpose, because then it just confuses people.

Currently we have `settings` module, which is the main source of configuration across Avocado. Is easily usable by users and frequently tweaked by them. When no config is provided, in-code defaults are chosen like default paths, runner-configuration, plugins configuration and about encoding we do have the `use-utf8` which is used by `varianter-mux-tree`.

So I'm against adding ad-hock `defaults` module, but if you think we reached the point where we need `defaults` module, create RFC, collect what fits into that category, add tools to find where the default comes from (currently we have nice `settings` handling where you learn where the values come from, adding `defaults` would make sense) etc.

Anyway now we have a regression, adding new defaults module does not sound like quick-fix especially near to the end of sprint. On the other hand we can revert the py3 handling, wait for the RFC and fix it properly (together with the `astring.ENCODING`).

Reply:
@ldoktor there are **no defaults settings whatsoever** defined in the `settings` module.  You're not seeing things clearly.

Currently the settings are in **configuration files**, not in the `settings` module.

The logic for a module such as `defaults` is so clear... suppose we want to load configuration files in a given encoding such as `utf-8`.  We need a place to define what is the encoding that is going to be used, and it can **not** be by using `settings` and a configuration file.

Reply:
This is low-level/internal defaults. Python 2.7 defines the default encoding as `ascii`. Python 3 changes that for `utf-8`. We don't allow users to change python default encoding from the avocado's config file in any case (what we allow is to set the encoding for the output purposes). As we have to keep Py2 and Py3 compatibility, apart from the benefit of having a default encoding for the project regardless the python version, setting that internally as a project default makes sense for me. This module seems the way to go. If we are doing the same thing (low level defaults) in the configuration file, it should be moved here.

Reply:
Depends on the view, currently we use `settings` as the main source of information. Some are in `/etc` but the real defaults come from code. This is the status quo. If we want to change it we should not do this ad-hock, but an RFC with some exploration would be required and more importantly most/all defaults should be evaluated. Otherwise we end-up with some using this solution, others using that solution and third adding new one just because they are not happy about the situation.

I understand the problem of loading the configuration (which is not part of this PR). Using the current status it'd mean hardcoding it in `settings` module. Yes, there are better solutions, one of them `defauls` module (as used in `avocado-vt`) another could be sharing it with the `astring.ENCODING` when added. Another could attempt to detect the encoding... there are plenty of solutions and they should be evaluated first. Anyway changing encoding of `settings` module is not the aim of this PR, the aim here is to fix regression until we get proper support for encoding (which is already in our current sprint backlog).

Reply:
@ldoktor don't mind the point about `settings` versus `defaults`, we've reached a majority in this particular discussion.

Reply:
Yes, as I said, it's one possible solution, but adding such module requires deeper thoughts and should be accompanied with systematical change in all affected defaults (I don't say all `settings` defaults, but those that makes sense being included in `defaults`) as well as some documentation and ideally support in `avocado config` to help users understanding the current structure.

I still don't think we should hurry and add such module now, but as you said, you got the majority and hopefully it'll be improved eventually. Or just forgotten and maintained for eons.

Reply:
For new concepts we also do create RFCs to avoid doing small steps into wrong (or just not best) directions.

Reply:
Do you honestly think a Python file containing one constant is a "new concept"? I think you're thinking too much of this... this is much more mundane code, with a simple and direct purpose.

Reply:
It's a new module which needs to be properly defined and already existing features that fit to it's purpose should be migrated, otherwise we'll have 2 different implementations of the same thing (in-code-defaults vs. `defaults.py`). Also using `defaults.py` is not the only solution so perhaps we might want to discuss (RFC) first the pros and cons as it affects the configuration, which is the base of the program. So yes, it's a new concept no matter how easy it looks.

PS: To the `config` file encoding, I don't think that users would like to modify avocado sources (`defaults.py`) to change default encoding. So ascii, hardcoded details, autodetection or argument would serve better.